docs/concepts-and-explanations/javatemplate.md

```diff
@@ -120,6 +120,51 @@ JavaTemplate.builder("Duration.ofMillis(#{any(int)})")
 There are many examples of `JavaTemplate` stubs in [rewrite-testing-frameworks](https://github.com/openrewrite/rewrite-testing-frameworks) and [rewrite-spring](https://github.com/openrewrite/rewrite-spring).
 :::
 
+## Context-free vs. Context-sensitive JavaTemplates
+
+There are two types of JavaTemplates – each having their own advantages/disadvantages:
+
+A **context-free** JavaTemplate does not depend on the surrounding LST. It is a static template that simply replaces or inserts code without considering the broader context of the source file. This is particularly useful for simple transformations like replacing method calls or inserting boilerplate code.
+
+For instance, the following template, when applied, would not check the existing imports or variables in scope:
+
+```java
+JavaTemplate.builder("System.out.println(\"Hello, World!\");")
+            .build();
+```
+
+The main advantage of context-free templates is that they only need to be parsed once before their parameters can be substituted. This has a much smaller performance impact on recipe runs.
+
+A **context-sensitive** JavaTemplate, on the other hand, is aware of and adapts to the surrounding LST. It can capture elements from the existing code, such as variable names, types, or method parameters.
+
+An example can help this make more sense. Let's presume we have code that looks like this:
+
+```java
+public void logMessage() {
+    String message = "Application started";
+    System.out.println(message);
+}
+```
+
+We want to change the `System.out.println` to be a logging method, but we don't want to use the variable `message`. We could write a template like:
+
+```java
+JavaTemplate.builder("logger.info(#{any(String)});")
+            .contextSensitive()
+            .build();
+```
+
+When applied, the template might generate:
+
+```java
+public void logMessage() {
+    String message = "Application started";
+    logger.info(message);
+}
+```
+
+This context of the surrounding code comes with a performance impact, though. That's not to say that you _shouldn't_ use them but, rather, that you should be aware of the cost of using them.
```

Comment:
Maybe we can here also explain what that cost is: Every time a context-sensitive JavaTemplate is applied, a small Java source file capturing all imports and declarations that are in scope needs to be generated and compiled using the Java compiler.

Reply:
Great suggestion! Updated.